rtl/eclrtl/rtldynfield.cpp

```diff
@@ -1405,202 +1381,516 @@ class GeneralRecordTranslator : public CInterfaceOf<IDynamicTransform>
+    {
+        return doTranslate(builder, callback, 0, fetcher);
+    }
+    virtual bool canTranslate() const override
+    {
+        return (matchFlags & match_fail) == 0;
+    }
+    virtual bool needsTranslate() const override
+    {
+        return (matchFlags & ~(match_link|match_inifblock)) != 0;
+    }
+    virtual bool needsNonVirtualTranslate() const override
+    {
+        return (matchFlags & ~(match_link|match_virtual|match_keychange|match_inifblock)) != 0;
+    }
+    virtual bool keyedTranslated() const override
+    {
+        return (matchFlags & match_keychange) != 0;
+    }
+protected:
+    virtual RecordTranslatorBase *createRecordTranslator(const RtlRecord &destRecInfo, const RtlRecord &srcRecInfo) const
+    {
+        CallbackRecordTranslator *ret = new CallbackRecordTranslator(destRecInfo, srcRecInfo);
+        ret->init(false);
+        return ret;
+    }
+    static size32_t translateScalarFromUtf8(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, const RtlTypeInfo &destType, const RtlTypeInfo &sourceType, const char *source, size_t srcLen)
+    {
+        switch(destType.getType())
+        {
+        case type_boolean:
+        case type_int:
+        case type_swapint:
+        case type_packedint:
+        case type_filepos:
+        case type_keyedint:
+        {
+            __int64 res = rtlUtf8ToInt(srcLen, source);
+            offset = destType.buildInt(builder, offset, field, res);
+            break;
+        }
+        case type_real:
+        {
+            double res = rtlStrToReal(srcLen, source);
+            offset = destType.buildReal(builder, offset, field, res);
+            break;
+        }
+        case type_data:
+        case type_string:
+            //Special case if source and destination types are identical to avoid cloning strings
+            if ((destType.fieldType & (RFTMkind|RFTMebcdic)) == sourceType.getType())
+            {
+                offset = destType.buildString(builder, offset, field, srcLen, source);
+                break;
             }
-            matchFlags |= info.matchType;
+            //fallthrough
+        case type_decimal:  // Go via string - not common enough to special-case
+        case type_varstring:
+        case type_qstring:
+        {
+            offset = destType.buildString(builder, offset, field, srcLen, source);
+            break;
         }
-        if (sourceRecInfo.getNumFields() > destRecInfo.getNumFields())
-            matchFlags |= match_remove;
-        if (matchFlags && !destRecInfo.getFixedSize())
+        case type_utf8:
+            //MORE: Could special case casting from utf8 to utf8 similar to strings above
+        case type_unicode:
+        case type_varunicode:
         {
-            for (unsigned idx = 0; idx < sourceRecInfo.getNumFields(); idx++)
+            size32_t utf8chars;
+            rtlDataAttr utf8Text;
+            rtlUtf8ToUtf8X(utf8chars, utf8Text.refstr(), srcLen, source);
```

Comment:
not needed - assignment is direct, same as all the other calls above.

Reply:
right, will change.